ntp.go

```diff
@@ -62,52 +71,82 @@ func (m *msg) SetMode(md mode) {
 	m.LiVnMode = (m.LiVnMode & 0xf8) | byte(md)
 }
 
+// SetTransmitTime sets the NTP protocol Transmit time
+func (m *msg) SetTransmitTime(t time.Time) {
+	m.TransmitTime = toNtpTime(t)
+}
+
 // Time returns the "receive time" from the remote NTP server
 // specifed as host.  NTP client mode is used.
-func getTime(host string, version byte) (time.Time, error) {
+func getTime(host string, version byte) (*msg, error) {
 	if version < 2 || version > 4 {
 		panic("ntp: invalid version number")
 	}
 
 	raddr, err := net.ResolveUDPAddr("udp", host+":123")
 	if err != nil {
-		return time.Now(), err
+		return nil, err
 	}
 
 	con, err := net.DialUDP("udp", nil, raddr)
 	if err != nil {
-		return time.Now(), err
+		return nil, err
 	}
 	defer con.Close()
 	con.SetDeadline(time.Now().Add(5 * time.Second))
 
 	m := new(msg)
 	m.SetMode(client)
 	m.SetVersion(version)
+	m.SetTransmitTime(time.Now())
 
 	err = binary.Write(con, binary.BigEndian, m)
 	if err != nil {
-		return time.Now(), err
+		return nil, err
 	}
 
 	err = binary.Read(con, binary.BigEndian, m)
 	if err != nil {
-		return time.Now(), err
+		return nil, err
 	}
 
-	t := m.ReceiveTime.UTC().Local()
-	return t, nil
+	return m, nil
 }
 
 // TimeV returns the "receive time" from the remote NTP server
 // specifed as host.  Use the NTP client mode with the requested
 // version number (2, 3, or 4).
 func TimeV(host string, version byte) (time.Time, error) {
-	return getTime(host, version)
+	m, err := getTime(host, version)
+	if err != nil {
+		return time.Now(), err
+	}
+	return m.ReceiveTime.UTC().Local(), nil
 }
 
 // Time returns the "receive time" from the remote NTP server
 // specifed as host.  NTP client mode version 4 is used.
 func Time(host string) (time.Time, error) {
-	return getTime(host, 4)
+	return TimeV(host, 4)
+}
+
+// Offset returns the offset in nanoseconds
+func Offset(host string) (int, error) {
+	m, err := getTime(host, 4)
+	if err != nil {
+		return -1, err
+	}
+	return offset(m.OriginTime.UTC(), m.ReceiveTime.UTC(), m.TransmitTime.UTC(), time.Now())
+}
+
+// offset variable names based off the rfc:
+// https://tools.ietf.org/html/rfc2030
+func offset(t1, t2, t3, t4 time.Time) (int, error) {
+	t1n := t1.UnixNano()
+	t2n := t2.UnixNano()
+	t3n := t3.UnixNano()
+	t4n := t4.UnixNano()
+	// https://tools.ietf.org/html/rfc2030 page 12
+	offset := ((t2n - t1n) + (t3n - t4n)) / 2
```

Comment:
Because I don't trust us to have considered all possible edge cases. If we use the rfc algo as intended (on ntp timestamps) then we don't need to worry about it, and can just follow it. It may also be worth digging into ntp sources to see how it's done there.